Data/Base.rte/Devices/Tools/GrappleGun/Scripts/RopeStateManager.lua

```diff
@@ -0,0 +1,622 @@
+end
+
+--[[
+  Checks for collisions when the grapple is flying, to transition to an attached state.
+  @param grappleInstance The grapple instance.
+  @return True if the state changed (grapple attached), false otherwise.
+]]
+function RopeStateManager.checkAttachmentCollisions(grappleInstance)
+	if grappleInstance.actionMode ~= 1 then 
+		Logger.debug("RopeStateManager.checkAttachmentCollisions() - Not in flying state (actionMode=%d), skipping", grappleInstance.actionMode)
+		return false 
+	end -- Only process in flying state.
+	
+	Logger.debug("RopeStateManager.checkAttachmentCollisions() - Starting collision detection")
+	
+	local stateChanged = false
+	
+	-- Much stricter collision detection with minimal ranges
+	local baseRayLength = math.max(1, (grappleInstance.Diameter or 4) * 0.2) -- Reduced from 0.5 to 0.2
+	local velocityComponent = math.min(1, (grappleInstance.Vel and grappleInstance.Vel.Magnitude or 0) * 0.1) -- Reduced from 0.2 to 0.1
+	local rayLength = baseRayLength + velocityComponent
+	rayLength = math.max(1, rayLength) -- Reduced minimum from 2 to 1
+
+	Logger.debug("RopeStateManager.checkAttachmentCollisions() - Ray parameters: baseLength=%.2f, velocityComponent=%.2f, finalLength=%.2f", 
+				 baseRayLength, velocityComponent, rayLength)
+
+	local rayDirection = Vector(1,0) -- Default direction
+	-- Require higher velocity threshold for directional casting
+	if grappleInstance.Vel and grappleInstance.Vel.Magnitude and grappleInstance.Vel.Magnitude > 0.1 then -- Increased from 0.005 to 0.1
+		local mag = grappleInstance.Vel.Magnitude
+		if mag ~= 0 then
+			rayDirection = Vector(grappleInstance.Vel.X / mag, grappleInstance.Vel.Y / mag)
+			Logger.debug("RopeStateManager.checkAttachmentCollisions() - Using velocity-based ray direction: (%.2f, %.2f), magnitude=%.2f", 
+						 rayDirection.X, rayDirection.Y, mag)
+		end
+	else
+		Logger.debug("RopeStateManager.checkAttachmentCollisions() - Using default ray direction (low velocity)")
+	end
+	
+	-- Primary ray (much shorter and more precise)
+	local collisionRay = rayDirection * rayLength
+	local hitPoint = Vector()
+	
+	-- Secondary ray (extremely short)  
+	local secondaryRayLength = math.max(0.5, baseRayLength * 0.1) -- Reduced from 0.3 to 0.1
+	local secondaryHitPoint = Vector()
+	
+	-- Close-range radius (extremely minimal)
+	local closeRangeRadius = math.max(0.5, (grappleInstance.Diameter or 4) * 0.1) -- Reduced from 0.3 to 0.1
+	local terrainHit = false
+
+	Logger.debug("RopeStateManager.checkAttachmentCollisions() - Secondary ray length: %.2f, close range radius: %.2f", 
+				 secondaryRayLength, closeRangeRadius)
+
+	-- 1. Check for Terrain Collision (primary ray) - require much higher strength
+	Logger.debug("RopeStateManager.checkAttachmentCollisions() - Performing primary terrain ray cast")
+	local terrainHit = SceneMan:CastStrengthRay(grappleInstance.Pos, collisionRay, 15, hitPoint, 0, rte.airID, grappleInstance.mapWrapsX) -- Increased from 5 to 15
```

Comment:
Produces this error sometimes, making the grapple gun just fire the grappler with broken rope and instantly reloading
```ERROR: ...te/Devices/Tools/GrappleGun/Scripts/RopeStateManager.lua:101: attempt to index upvalue 'rte' (a nil value)```